federationtypes.go

```diff
@@ -429,27 +429,42 @@ func (r RespSendJoin) Check(ctx context.Context, keyRing JSONVerifier, joinEvent
 		return err
 	}
 
-	stateEventsByID := map[string]*Event{}
-	authEvents := NewAuthEvents(nil)
+	eventsByID := map[string]*Event{}
+	stateEvents := NewAuthEvents(nil)
+
+	// Since checkAllowedByAuthEvents needs to be able to look up any of the
+	// auth events by ID only, we will build a map which contains references
+	// to all of the auth events.
+	for i, event := range r.AuthEvents {
+		eventsByID[event.EventID()] = &r.AuthEvents[i]
+	}
+
+	// Then we add the current state events too, since our newly formed
+	// membership event will likely refer to these as auth events too.
 	for i, event := range r.StateEvents {
-		stateEventsByID[event.EventID()] = &r.StateEvents[i]
-		if err := authEvents.AddEvent(&r.StateEvents[i]); err != nil {
+		eventsByID[event.EventID()] = &r.StateEvents[i]
```

Comment:
Isn't there a risk that this will set the same event twice? I suppose that's probably not a problem?

Reply:
In room versions 1 and 2 it is possible and that might break the join, but that's somewhat unavoidable. In room versions 3 and onwards, it is no longer possible because by changing the event—and by extension the reference hash—we inevitably change the event ID.